File: ios/CustomMessageHandlerBridge.swift
import BitmovinPlayer
import ExpoModulesCore

internal class CustomMessageHandlerBridge: NSObject {
#if os(iOS)
    private(set) lazy var customMessageHandler: CustomMessageHandler = {
        let customMessageHandler = CustomMessageHandler()
        customMessageHandler.delegate = self
        return customMessageHandler
    }()
#endif

    private let nativeId: NativeId
    private weak var delegate: CustomMessageHandlerBridgeDelegate?

    init(_ nativeId: NativeId, delegate: CustomMessageHandlerBridgeDelegate?) {
        self.nativeId = nativeId
        self.delegate = delegate
    }

    func receivedSynchronousMessage(_ message: String, withData data: String?) -> String? {
        delegate?.receivedSynchronousMessage(
            nativeId: nativeId,
            message: message,
            withData: data
        )
    }

    func receivedAsynchronousMessage(_ message: String, withData data: String?) {
        delegate?.receivedAsynchronousMessage(
            nativeId: nativeId,
            message: message,
            withData: data
        )
    }

    func sendMessage(_ message: String, withData data: String?) {
#if os(iOS)
        customMessageHandler.sendMessage(message, withData: data)
#endif
    }
}

#if os(iOS)
extension CustomMessageHandlerBridge: CustomMessageHandlerDelegate {}
#endif

internal protocol CustomMessageHandlerBridgeDelegate: AnyObject {
    func receivedSynchronousMessage(nativeId: NativeId, message: String, withData data: String?) -> String?
    func receivedAsynchronousMessage(nativeId: NativeId, message: String, withData data: String?)
}

File: android/src/main/java/com/bitmovin/player/reactnative/ui/CustomMessageHandlerBridge.kt
package com.bitmovin.player.reactnative.ui

import android.webkit.JavascriptInterface
import com.bitmovin.player.reactnative.CustomMessageHandlerModule
import com.bitmovin.player.reactnative.NativeId
import com.bitmovin.player.ui.CustomMessageHandler

class CustomMessageHandlerBridge(
    private val nativeId: NativeId,
    private val module: CustomMessageHandlerModule? = null,
) {
    val customMessageHandler = CustomMessageHandler(
        object : Any() {
            @JavascriptInterface
            fun sendSynchronous(
                name: String,
                data: String?,
            ): String? = module?.receivedSynchronousMessage(nativeId, name, data)

            @JavascriptInterface
            fun sendAsynchronous(
                name: String,
                data: String?,
            ) = module?.receivedAsynchronousMessage(nativeId, name, data)
        },
    )

    fun sendMessage(message: String, data: String?) = customMessageHandler.sendMessage(message, data)
}
